src/garage/tf/algos/te_npo.py

```diff
@@ -0,0 +1,1191 @@
+
+        tasks = [path['tasks'] for path in paths]
+        tasks = pad_tensor_n(tasks, max_path_length)
+
+        latents = [path['latents'] for path in paths]
+        latents = pad_tensor_n(latents, max_path_length)
+
+        rewards = [path['rewards'] for path in paths]
+        rewards = pad_tensor_n(rewards, max_path_length)
+
+        returns = [path['returns'] for path in paths]
+        returns = pad_tensor_n(returns, max_path_length)
+
+        baselines = pad_tensor_n(baselines, max_path_length)
+
+        trajectories = np.stack([path['trajectories'] for path in paths])
+
+        agent_infos = [path['agent_infos'] for path in paths]
+        agent_infos = stack_tensor_dict_list(
+            [pad_tensor_dict(p, max_path_length) for p in agent_infos])
+
+        latent_infos = [path['latent_infos'] for path in paths]
+        latent_infos = stack_tensor_dict_list(
+            [pad_tensor_dict(p, max_path_length) for p in latent_infos])
+
+        trajectory_infos = [path['trajectory_infos'] for path in paths]
+        trajectory_infos = stack_tensor_dict_list(
+            [pad_tensor_dict(p, max_path_length) for p in trajectory_infos])
+
+        env_infos = [path['env_infos'] for path in paths]
+        env_infos = stack_tensor_dict_list(
+            [pad_tensor_dict(p, max_path_length) for p in env_infos])
+
+        valids = [np.ones_like(path['returns']) for path in paths]
+        valids = pad_tensor_n(valids, max_path_length)
+
+        lengths = np.asarray([v.sum() for v in valids])
+
+        samples_data = dict(
+            observations=obs,
+            actions=actions,
+            tasks=tasks,
+            latents=latents,
+            trajectories=trajectories,
+            rewards=rewards,
+            baselines=baselines,
+            returns=returns,
+            valids=valids,
+            lengths=lengths,
+            agent_infos=agent_infos,
+            env_infos=env_infos,
+            latent_infos=latent_infos,
+            trajectory_infos=trajectory_infos,
+            paths=paths,
+            average_return=np.mean(undiscounted_returns),
+        )
+
+        return samples_data
+
+    def _build_optimizer(self, optimizer, optimizer_args):
+        """Build up optimizer for policy.
+
+        Args:
+            optimizer (obj): Policy optimizer. Should be one of the optimizers
+                in garage.tf.optimizers.
+            optimizer_args (dict): The arguments of the optimizer.
+
+        Returns:
+            obj: Policy optimizer. Should be one of the optimizers
+                in garage.tf.optimizers.
+            dict: The arguments of the optimizer.
+
+        """
+        # pylint: disable=no-self-use
+        if optimizer is None:
+            optimizer = LbfgsOptimizer
+        if optimizer_args is None:
+            optimizer_args = dict()
+        return optimizer, optimizer_args
+
+    def _build_inference_optimizer(self, optimizer, optimizer_args):
```

Comment:
same question -- can't this be a function?

also, this function seems superfluous -- can't you just use build_optimizer twice?